### PR TITLE
IP: Undo split indices for sensitive text detection

### PR DESCRIPTION
**Pull request by sarayourfriend**
## Fixes

Part of https://github.com/WordPress/openverse/issues/3336 by @AetherUnbound 

## Description

This discussion is following the Openverse decision-making process. Information about this process can be found [on the Openverse documentation site](https://docs.openverse.org/meta/decision_making/index.html). Requested reviewers or participants will be following this process. If you are being asked to give input on a specific detail, you do not need to familiarise yourself with the process and follow it.

### Current round

This discussion is currently in the **Decision round**.

The deadline for review of this round is **2024-09-25**.

## Checklist

- [x] My pull request has a descriptive title (not a vague title like`Update index.md`).
- [x] My pull request targets the _default_ branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [N/A] I added or updated tests for the changes I made (if applicable).
- [N/A] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.
- [N/A] I ran the DAG documentation generator (`ov just catalog/generate-docs` for catalog
      PRs) or the media properties generator (`ov just catalog/generate-docs media-props`
      for the catalog or `ov just api/generate-docs` for the API) where applicable.

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: documentation/projects/proposals/trust_and_safety/detecting_sensitive_textual_content/20240903-implementation_plan_undo_split_filtered_indices.md [hunk 1]
```diff
@@ -0,0 +1,830 @@
...
+thus increasing its disk operations.
+
+To actually implement a simultaneous comparison, I hypothesise that we would
+need to have identically resourced clusters, one for executing the new type of
+query, and one for the old type. I believe because the performance question in
+this plan has to do with reducing disk operations and increasing cache
+utilisation, that the metrics are unique from relevance optimisations, which we
+would be able to measure side by side.
+
+While possible in theory, it is too complex!
```
Comment (dhruvkb) [inline documentation/projects/proposals/trust_and_safety/detecting_sensitive_textual_content/20240903-implementation_plan_undo_split_filtered_indices.md:134]: One approach that we could take is to qualitatively estimate if the single index approach makes things better is to vary the portion of queries that use the new index. 

If our hypothesis is correct, increasing the portion of queries using the new index should correspond to a in overall resource usage. This way we also don't have to drop the feature flag outright and we can gradually switch over 100% of requests to the new index.

File: documentation/projects/proposals/trust_and_safety/detecting_sensitive_textual_content/20240903-implementation_plan_undo_split_filtered_indices.md [hunk 2]
```diff
@@ -0,0 +1,830 @@
...
+            },
+        },
+    }
+```
+
+The API unit tests run against the real Elasticsearch cluster, so the result of
+running the script as part of the sensitive-media document update should be
+"painless"[^😛] to test.
+
+[^😛]: 😛
```
Comment (dhruvkb) [inline documentation/projects/proposals/trust_and_safety/detecting_sensitive_textual_content/20240903-implementation_plan_undo_split_filtered_indices.md:589]: 😛

**Review by dhruvkb — APPROVED** [review]
The plan looks good to me. The steps are logical, the changes to the API look correct and the approximate analysis of the performance impact also makes sense.

File: documentation/projects/proposals/trust_and_safety/detecting_sensitive_textual_content/20240903-implementation_plan_undo_split_filtered_indices.md (hunk 1 above)
Comment (sarayourfriend) [reply documentation/projects/proposals/trust_and_safety/detecting_sensitive_textual_content/20240903-implementation_plan_undo_split_filtered_indices.md:134]: True, it would be cool if we had methods for doing that! We don't have anything currently set up that let us adjust settings on the fly like that. It would be interesting if we had a Django model for settings like that where we might want more responsive control than environment variables provide. I anticipate something like that would come from #421... I'd be wary of implementing anything like that for this, and doing several deployments with small environment variable changes each time would be a headache compared to just doing a clean switch over and rolling back if needed.

> If our hypothesis is correct, increasing the portion of queries using the new index should correspond to a in overall resource usage

One part of this that I'm not confident in is... how many queries would need to be redirect to the new method for it to make a difference for Elasticsearch? The resource consumption comes from the fact that ES has to pull results from two queries instead of one (and sometimes those raw documents are identical, so it's wasting cache at least some of the time).

In this particular case, it's such a radically different approach between the two query methods (one index vs two) that I don't know what to anticipate from Elasticsearch if it were getting both. Other than that I'm fairly confident it would worsen caching because it would be getting more "types" of queries to the unfiltered index. If at some point of switching to more "new" queries it ever does improve cacheability, the percentage of "old" style queries would have to be very low, I would think. How ES responds to the frequency, and whether we would start to see improvements in cache performance over time, are a question. My guess is never, so long as both indices are being used, it will continue to require pulling from both indices on disk, and end up using a lot of its cache to do so.

That is really only a guess. My point is that I just don't know enough about how ES caches to say that we would see a corresponding change by moving the search method over slowly.

---

That being said, it would be cool if we had more responsive controls and things like what you're saying were more easily possible. I could see it being very useful for performance testing new query functions, for example. If we could ramp up, rather than cut over immediately, and not need to have a bunch of individual deployments for env variable changes, we'd probably get a better picture of how things compare under certain conditions. I head someone from Instagram talk about doing this with traffic to Instagram deployments running different modifications of CPython (part of their fast-python stuff), where they slowly move a percentage of traffic over and get a read on resource usage. Very cool! Must be lots of interesting dev tooling to make it nice to do on a regular basis.

**Comment by sarayourfriend** [comment]
> Could you define specific prerequisites for the "cleanup" steps?

Sure thing, good call out. When I get to revision (after Staci reviews for clarification round), I'll add something like the following:

> Clean-up should occur only after 2 weeks of running the new approach in production, including two full production data refreshes. This is to ensure sufficiently exericse the new approach both during the data refresh and at query time before starting to take actions that will make rolling back much more cumbersome.

Does that sound alright?

**Review by stacimc — COMMENTED** [review]
This looks great to me, @sarayourfriend -- I had a question about the indexer worker in the local dev environment, but that should be easily handled. I'm curious about your thoughts on the ingestion approach, but I think this approach will work well and I see the tradeoffs.

File: documentation/projects/proposals/trust_and_safety/detecting_sensitive_textual_content/20240903-implementation_plan_undo_split_filtered_indices.md [hunk 3]
```diff
@@ -0,0 +1,830 @@
...
+final comment, whilst dedicated ingest nodes are not required to work with
+ingest pipelines, a cluster still needs at least one node with the ingest role
+to run ingest pipelines (it just isn't required to be those nodes' only role).
+In order to service that requirement, we would need to introduce a new
+complexity into our Elasticsearch cluster. We do not have the capacity (time and
+knowledge) to fulfil this requirement. Therefore, we should not use
+Elasticsearch ingest pipelines to generate the sensitivity designation.
+
+### Move sensitive text detection earlier in the data refresh pipeline
+
```
[the remark this answers is not in the document]
Comment (sarayourfriend) [reply documentation/projects/proposals/trust_and_safety/detecting_sensitive_textual_content/20240903-implementation_plan_undo_split_filtered_indices.md:711]: I did consider that, and for all the money in the world I would have sworn I'd written about it in the IP, but I have been thinking of [this comment of mine from the issue](https://github.com/WordPress/openverse/issues/3336#issuecomment-2288432417). It definitely warrants covering in the IP directly.

The things I like about it are the ones you shared, and that we can avoid needlessly re-running the check on works.

The things that make me nervous about it, which made me not want to pursue it, are:

1. A lack of consistency guaranteed in a single data refresh cycle if we are updating sensitivity. If a data refresh runs during the update, then it will have records with indications from the previous list and others from the new list. We could block data refreshes from occurring if a sensitivity update needs to happen (when we change the terms list) to prevent that. Assuming it's possible to do a full sensitivity update in <24 hours (assumed because our current processes are able to accomplish that) then it isn't that big a deal, it would delay the data refresh by that amount of time.
2. Uncertainty about whether we would need two different implementations of sensitive text detection. Our ingestion happens in Python, but batched updates in SQL. As far as I know, we don't have good ways to write parts of ingestion in SQL, and no way to write any part of batched updates in Python (due to RDS's lack of support for ulpython). So my assumption is that we would need a version in SQL to account for the batched update version, and then another in Python for ingestion of new records. However, it occurred to me now that we could use a PGSQL function for this, and then use an insert trigger to derive the `sensitive_text` sensitivity, and the same function could be referenced in the batched update. In other words, I think it may be possible to create a single implementation that accounts for both.
  - My primary concern with two separate implementations is in one not being used often enough for us to be sure that it's "still working". We have not made many changes to the sensitive terms list (I think maybe once, if ever, since implementing its usage). If the SQL version used for the batched update doesn't get regularly used, we risk an unfortunate surprise of it being broken when we make a change to the terms. To fully test this locally, I think we'd need to use the full terms list, or else try to identify all the edge cases (emoji, unicode, special characters, etc), which will need special handling/escaping for put into SQL. Not exercising that regularly is a big risk, long term. The consequence is the surprise of it not working when we most expect and rely on it to work.
3. Uncertainty about the performance impact on the scheduler which is also our executor, due to local executor usage. This is only relevant if the implementation is in Python, which as I've written in the previous point, may not be necessary. So, this is less of a concern to me, as I feel comfortable offloading this kind of work to PG.

I'm actually curious to look more into this. @stacimc what do you think about these points? Do they fit with your understanding of the situation around a popularity-refresh-like approach, _viz_ the limitations and approach we would need to take? In particular, do you have an opinion or sense about point 2, and whether we would be able to have a single implementation rather than a SQL and Python one separately? A single implementation and just accepting a delay to one data refresh using our existing concurrency controls would resolve the questions I had about it. In which case, I think it's undeniably the better approach.

**Comment by sarayourfriend** [comment]
I've also been thinking about this IP for the last week and regretting my recommendation of the `sensitivity` list. I think instead,  an object of boolean properties like `sensitivity: { text: boolean, user_reported: boolean }` would be better. It could also have an `any: boolean` field as a normalised version of all the booleans in the object, which we could query against for simpler non-sensitive queries, which are the predominant kind of queries we make. Regardless of the version we go with, I want to change the IP to go with this approach. It has the following advantages:

1. It does not require using a Painless script to update the document in the index (way simpler, less fiddly, easier to test and maintain).
2. It produces an identical type of `must_not` term query on the boolean property/ies, matching our current query. Therefore, we can be confident it will have identical performance characteristics to our current query.
3. It is potentially more flexible long term, as an object can much more naturally grow in properties to include things like which fields had sensitive text, which sensitive terms were detected, and so forth. Things that may be valuable or even essential information to improving how our sensitive text detection works in the future.

The last advantage particularly applies in the context of a catalogue-based approach like the one Staci asked about [in this comment](https://github.com/WordPress/openverse/pull/4904#discussion_r1759200541).

File: documentation/projects/proposals/trust_and_safety/detecting_sensitive_textual_content/20240903-implementation_plan_undo_split_filtered_indices.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (stacimc) [reply documentation/projects/proposals/trust_and_safety/detecting_sensitive_textual_content/20240903-implementation_plan_undo_split_filtered_indices.md:711]: To point 1, exactly, it would be very easy to add the `recalculate_sensitivity` DAG to the list of blocking DAGs for the data refresh, and that should be fine _as long as recalculating sensitivity doesn't take too long_ like you said. We'd expect this to happen infrequently, as well.

To point 2, we actually can do SQL updates during ingestion by adding a [CalculatedColumn](https://teamopenverse.wordpress.com/2024/07/25/elasticache-service-update/); this column is not populated by the provider scripts, but given a SQL function that runs during the [upsert query](https://github.com/WordPress/openverse/blob/a4b1bf4e121373d98a2a81bc38a4556a86c05695/catalog/dags/common/loader/sql.py#L316). This is what we do with standardized popularity; the provider ingesters don't "know about" popularity at all. Then the same SQL function is used for the batched popularity updates.

My biggest question is about performance as well, because a full batched update of all records is quite time consuming. Instead I think we should SELECT only the records that _need_ to be updated after an update to the terms list -- ie, all records that match any of the sensitive terms that were either deleted or added. I'm not sure how performant _that_ query would be, that would be interesting to look into.

[the remark this answers is not in the document]
Comment (sarayourfriend) [reply documentation/projects/proposals/trust_and_safety/detecting_sensitive_textual_content/20240903-implementation_plan_undo_split_filtered_indices.md:711]: > To point 2, we actually can do SQL updates during ingestion by adding a [CalculatedColumn](https://teamopenverse.wordpress.com/2024/07/25/elasticache-service-update/); this column is not populated by the provider scripts, but given a SQL function that runs during the [upsert query](https://github.com/WordPress/openverse/blob/a4b1bf4e121373d98a2a81bc38a4556a86c05695/catalog/dags/common/loader/sql.py#L316). This is what we do with standardized popularity; the provider ingesters don't "know about" popularity at all. Then the same SQL function is used for the batched popularity updates.

Perfect!

> My biggest question is about performance as well, because a full batched update of all records is quite time consuming. Instead I think we should SELECT only the records that need to be updated after an update to the terms list -- ie, all records that match any of the sensitive terms that were either deleted or added. I'm not sure how performant that query would be, that would be interesting to look into.

Just to make sure I understand your suggestion correctly, are you saying that we would use the `where_query` to `SELECT * FROM image WHERE check_for_sensitive_terms((title, description, tags->name), <new terms>) OR check_for_sensitive_terms(description, <new terms>) OR check_for_sensitive_terms(tags->name, <new terms>)` (where `check_for_sensitive_terms` is the function used in all sensitive term checks including on insert)?

<small>As a bit of an aside, this is actually an interesting question to me that I feel I don't yet understand. Is it an optimisation to filter the set of records beforehand, rather than doing a full update of the table? To my mind, the number of invocations of `check_for_sensitive_terms` would be identical in each, so it comes down to whether PG is happier doing that as part of a full table scan in the select rather than in the individual update batches. Either way, the performance _viz_ the check is O(n) where `n` is the number of rows. If there is a different way to view this, it would be helpful for me to incorporate into my understanding of how to approach batched updates where the update is conditional on a derived quantity (so not simple equality, for example).</small>

Generally, though, it should be easy to test this. For one, we'd need to run this same select at least one time to update based on the current sensitive terms list. We can verify the viability of the `select` pretty easily, in that case, by manually running that query with a constructed sensitive terms list.

<details>
<summary>This function almost works to do it using PL/pgSQL... the terms list is just "water" and "running", but I've tested it with the regular terms list and it falls over on emojis and things like the <code>@</code> symbol.</summary>

```sql
CREATE OR REPLACE FUNCTION check_sensitive_terms(fields text[]) RETURNS BOOLEAN AS $$
  DECLARE
    sensitive_terms_list text;
    sensitive_terms_regexps text[];
    found_match boolean;
    term text;
    term_regexp text;
    field text;
  BEGIN
    sensitive_terms_list := encode(decode(
      'CndhdGVyCnJ1bm5pbmcKCg==',
      'base64'
    ), 'escape');

    sensitive_terms_regexps := ARRAY[]::text[];

    FOREACH term IN ARRAY regexp_split_to_array(sensitive_terms_list, '\n')
    LOOP
      IF term <> '' THEN
        sensitive_terms_regexps := sensitive_terms_regexps || format('\m%s\M', term);
      END IF;
    END LOOP;

    found_match := FALSE;

    <<fields_loop>>
    FOREACH field IN ARRAY fields
    LOOP

      FOREACH term_regexp IN ARRAY sensitive_terms_regexps
      LOOP
        found_match := field ~* term_regexp;
        EXIT fields_loop WHEN found_match;
      END LOOP;

    END LOOP;

    RETURN found_match;
  END;
$$
LANGUAGE plpgsql;
```

</details>

I bet Perl might be the better option for this. In any case, it's at least _possible_ with a function of this sort, even if there are problems we need to account for and fix.

Of course, there's also the glaring issue of parsing the terms and building the regex list on every call to the function. I'm going to guess the "right" way to do this would be to have a table of the regex patterns that the function queries from or is required in the argument list.

Maybe we can discuss this tomorrow during the team meeting and come up with a decision of whether we feel this is sufficiently doable that we can move forward with this as the plan, instead of doing it in the ingestion worker. I feel confident we can come up with _something_ that will work, even if we run into some snags.

[the remark this answers is not in the document]
Comment (sarayourfriend) [reply documentation/projects/proposals/trust_and_safety/detecting_sensitive_textual_content/20240903-implementation_plan_undo_split_filtered_indices.md:711]: @stacimc and I spent about 90 minutes or so today talking about this. We sorted out what seemed to be all the broad and many of the specific details required if we put the sensitive text detection into ingestion rather than in the data refresh.

Ultimately, while we came up with an idea that would work, it was _dramatically_ more complicated than this approach, and especially became complicated when trying to consider and account for the various race conditions that could occur with ingestion DAGs. It was all plausible, by our measure, but much more complex than the indexer worker version.

Therefore, (Staci, please correct me if I am wrong in this conclusion) we will stick with the indexer worker based approach as discussed in the current version of the IP, with the modifications I mentioned [in this comment](https://github.com/WordPress/openverse/pull/4904#issuecomment-2351839786).

I will also update the IP to include an explicit mention of this discussion and, without too much detail (there is a lot), note that the complexity outweighed any conceivable benefits we came up with, of which all were future theoretical benefits, rather than immediate ones.

**Comment by sarayourfriend** [comment]
@zackkrida I've added details for a cool-off period in the plan in this commit: https://github.com/WordPress/openverse/pull/4904/commits/13c2bebf36398eab10f3f309703ad7b7a6a2ad5a

@stacimc I've added details about the discussion we had yesterday re: moving the check into Airflow in this commit: https://github.com/WordPress/openverse/pull/4904/commits/27b3781775231005ce3eddd6cff2c5d21afbdad8

That second commit also includes the update to use a sensitivity object with boolean properties and a denormalised `any`.

I am waiting on one last question to clarify from @stacimc that I sent in Slack regarding how the indexer workers are used, and then I will be able to make a small change (it will be small either way) to address [this clarification Staci mentioned regarding the ephemerality of the indexer workers](https://github.com/WordPress/openverse/pull/4904#discussion_r1759186065).

File: documentation/projects/proposals/trust_and_safety/detecting_sensitive_textual_content/20240903-implementation_plan_undo_split_filtered_indices.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (sarayourfriend) [reply documentation/projects/proposals/trust_and_safety/detecting_sensitive_textual_content/20240903-implementation_plan_undo_split_filtered_indices.md:552]: Details for this are in this commit: https://github.com/WordPress/openverse/pull/4904/commits/27b3781775231005ce3eddd6cff2c5d21afbdad8

File: documentation/projects/proposals/trust_and_safety/detecting_sensitive_textual_content/20240903-implementation_plan_undo_split_filtered_indices.md [hunk 4]
```diff
@@ -0,0 +1,830 @@
...
+- Indexer worker field generation:
+  - Retrieve the sensitive terms list from the GitHub URL (see data refresh for
+    reference) and make the list available to the indexer processes without
+    requiring subsequent outbound requests for the list.
+    - Create a new Python module in the indexer worker, `sensitive_terms.py`. It
+      will export the following:
+      - A constant `SENSITIVE_TERMS_LOC` set to
+        `Path(__file__).parent / "sensitive_terms.txt"`. Indexer workers are
+        ephemeral, so it is safe to write to disk without risking accidental
+        reuse across runs.
```
[the remark this answers is not in the document]
Comment (sarayourfriend) [reply documentation/projects/proposals/trust_and_safety/detecting_sensitive_textual_content/20240903-implementation_plan_undo_split_filtered_indices.md:288]: I've pinged Staci in Slack to clarify one small detail about the indexer worker usage, and then I will be able to make a revision to clarify this point.

If the indexer workers manage at most one task, then we can simply move the call to `retrieve_sensitive_terms` into the handler for `POST /task`, and do so in all cases. If the indexer workers can receive multiple calls to `POST /task` in a single "lifecycle" (1 data refresh), then we do the call in api start up as already specified, but also call the function again in the `POST /task` handler whenever `ENVIRONMENT != "production"`.

**Comment by sarayourfriend** [comment]
@dhruvkb and @stacimc this is ready for y'all to take another look and make a decision or raise blockers. @dhruvkb I know you left an approval before but just wanted to wait until the decision round to lock it in so feel free to change your mind, of course! :slightly_smiling_face:

File: documentation/projects/proposals/trust_and_safety/detecting_sensitive_textual_content/20240903-implementation_plan_undo_split_filtered_indices.md [hunk 5]
```diff
@@ -0,0 +1,813 @@
...
+<!-- See the implementation plan guide for more information: https://github.com/WordPress/openverse/tree/19791f51c063d0979112f4b9f4eeace04c8cf5ff/docs/projects#implementation-plans-status-in-rfc -->
+<!-- This template is exhaustive and may include sections which aren't relevant to your project. Feel free to remove any sections which would not be useful to have. -->
+
+## Reviewers
+
+<!-- Choose two people at your discretion who make sense to review this based on their existing expertise. Check in to make sure folks aren't currently reviewing more than one other proposal or RFC. -->
+
+- [ ] @dhruvkb (API familiarity, especially in code that interacts with ES
+      during moderation)
+- [ ] @stacimc (Data refresh and ingestion worker expertise)
```
Comment (stacimc) [inline documentation/projects/proposals/trust_and_safety/detecting_sensitive_textual_content/20240903-implementation_plan_undo_split_filtered_indices.md:14]:
```suggestion
- [x] @stacimc (Data refresh and ingestion worker expertise)
```

**Review by stacimc — APPROVED** [review]
Fantastic! I love the addition of the denormalized `any` as well 👍

Thanks for indulging my questions about the approach -- and for so clearly documenting what ended up being a very complex discussion! 😄 Especially as we talk about the many different places data transformation is happening in our pipelines, it's so nice to really nail down our priorities and consider the options thoroughly. I feel very confident with the approach you've outlined here, cheers!

File: documentation/projects/proposals/trust_and_safety/detecting_sensitive_textual_content/20240903-implementation_plan_undo_split_filtered_indices.md [hunk 6]
```diff
@@ -0,0 +1,813 @@
...
+**Author**: @sarayourfriend
+
+<!-- See the implementation plan guide for more information: https://github.com/WordPress/openverse/tree/19791f51c063d0979112f4b9f4eeace04c8cf5ff/docs/projects#implementation-plans-status-in-rfc -->
+<!-- This template is exhaustive and may include sections which aren't relevant to your project. Feel free to remove any sections which would not be useful to have. -->
+
+## Reviewers
+
+<!-- Choose two people at your discretion who make sense to review this based on their existing expertise. Check in to make sure folks aren't currently reviewing more than one other proposal or RFC. -->
+
+- [ ] @dhruvkb (API familiarity, especially in code that interacts with ES
```
Comment (sarayourfriend) [inline documentation/projects/proposals/trust_and_safety/detecting_sensitive_textual_content/20240903-implementation_plan_undo_split_filtered_indices.md:12]:
```suggestion
- [x] @dhruvkb (API familiarity, especially in code that interacts with ES
```

**Comment by sarayourfriend** [comment]
This is past the deadline, and I've pinged in Slack with no response, so I'm going to merge based on Dhruv's previous PR review with the approval.

Thanks for the reviews, y'all.

**Review by krysal — APPROVED** [review]
I finally got to reply to this. Fantastic write-up, @sarayourfriend! Everything seems evaluated and well explained. Compared to the alternatives, the selected approach sounds wonderfully simple (not necessarily easy). I'm eager to try it, but I want to refrain from interfering since you have expressed the intention of continuing with it.

I left minor comments that don't block anything. Would you like me to merge it as is? Do you have the list of issues for a milestone?

**Comment by sarayourfriend** [comment]
@krysal please feel free to make any edits you'd like directly to the IP and merge it as you'd prefer. I can assist in the work if you'd like, otherwise, please go ahead and implement it however you see fit :+1:
